docs/specs/lcb-signals-and-events.md

```diff
@@ -0,0 +1,222 @@
+
+1. Signal emission and handling is not very well checked.  At a language level, there is no way for widgets to describe what type of signals they can emit.  Similarly, there is no syntax for distinguishing between widget handlers and "normal" handlers, meaning that (in principal) a handler that is not intended to be a signal handler could end up getting called in a signal context if a child widget happens to start emitting a signal with a matching name.  Finally, there is no facility for checking that signal prototypes match signal handler prototypes.
+
+2. Similar issues to 1. above apply for event handling.
+
+3. In LCS, event handling propagates upward: if a button doesn't have an `MouseUp` handler, then the event gets passed on to a parent card, and so on.  Furthermore, there is the idea of explicit propagation: it is possible to do something in an event handler and then, for example, `pass MouseUp` to invoke the normal event propagation logic.
+
+   In LCB, event propagation is *not* currently automatic.  If a widget does not explicitly handle and re-`post` each possible event, then its script (or parent widget) never sees the event.
+
+4. Sometimes, it's important to be able to be able to prevent child widgets from seeing an event, or performing some pre-processing / post-processing around the normal event handling process.  The current LCB event handling scheme does not permit this.
+
+## Assumptions
+
+For the purposes of this document we assume that:
+
+* All connection of signal and event handlers is implicit, rather than explicit.
+
+* An instance of a widget as an engine control can be considered as a composed widget, comprising a script and the "real" widget instance (this makes sense when you consider the fact that the widget control exists, has a script, and can be manipulated even if the "real" widget was unable to be loaded for some reason).  This will be referred to as a "host widget" in the rest of this document.
+
+## Signals
+
+### Signal emission
+
+A widget's LCB source code must explicitly name, and provide prototypes for, the signals that it can emit [1]:
+
+    public signal <name:id>(<params:formals>)
+
+Signal declarations must be placed in module scope.  The signal **name** must be unique in the module.
+
+Signals are explicitly emitted by a widget using an `emit` statement:
+
+    emit <name:id>(<params:expressions>)
+
+It is a run-time error to emit a signal with the wrong number or types of emission parameters [2].  The compiler may statically verify that signals are emitted with compatible parameter lists.
+
+The `emit` statement may return before all the signal handlers for the signal have finished running [3].
+
+The signal handlers for a signal get run exactly once for each emission.  All of the handlers for an emission must complete before any of the handlers for a subsequent emission begin [4].
+
+The signal handlers for a signal may be run in any order.
+
```

Comment:
@runrevmark said:
> Okay - so your use-case is one where all the data is in memory, but needs processing to turn it into a form for the view to display; thus you use a callback to format data rather than to fetch.

@montegoulding said:
> I'm a little confused here. If the data is already cached is the only reason you aren't setting the data in one hit because you don't want to duplicate it in memory?

Both of these statements get to the heart of the issue. The data is already in memory and portions of it need to be displayed to the user. While I currently use dispatch to get data to display, I realize that probably isn't the ideal way to handle things going forward with widgets. I don't think we have everything in place to solve this problem the right way though.

As an extreme (real world) example, think of a SQLite database cursor with 50,000 records. The view only needs to extract data from the records in the cursor that are currently visible (or just out of view). Trying to read everything into the current DataGrid in LiveCode is horrendously slow. Just displaying what is visible makes the DataView really snappy.

Another example is a document object that has a visual representation. As changes are made to the document you don't want to have to update the data in the dataview and update the data in the actual model.

> If so then perhaps that's an angle of attack and perhaps that's where you are going with your data provider module which might enable both widget data views and LCS to manipulate a common data model. Like I said before though I think something like this if done right needs abstract modules/libraries or protocols so you can have many data providers that use a common api. There's a few things that need to be in place for that to be a good option though (LCB database access etc).

I think you are right. It may just be that until we can get to a common data provider api that widgets only get data by setting a property. I've been trying to figure out how to get widgets to play nicely with data in LCS but there may not be a good approach for that just yet. Coming back to the original issue, making synchronous signals so that a widget can pull in row data in order to render it probably isn't the answer.